File: Net/SSOExtension/VersionOne.SDK.APIClient.SSOExtension/PostData.cs
using System;
using System.Collections;
using System.Web;

namespace VersionOne.SDK.APIClient.SSOExtension
{
    /// <summary>
    /// Data posted to SSO Providers (IDP and SP)
    /// This object needs to be created and returned by the class that implements IResponseProvider
    /// </summary>
    public abstract class PostData
    {
        private readonly ArrayList _formData = new ArrayList();

        /// <summary>
        /// Add a key-value pair
        /// results in post data that read key=value
        /// </summary>
        /// <param name="key"></param>
        /// <param name="value"></param>
        public void Add(string key, string value)
        {
            _formData.Add(string.Format("{0}={1}", key, HttpUtility.UrlEncode(value)));
        }

        /// <summary>
        /// adds a key with no value
        /// results in post data that reads key=
        /// </summary>
        /// <param name="key"></param>
        public void Add(string key)
        {
            _formData.Add(string.Format("{0}=", key));
        }

        /// <summary>
        /// Returns the post data string
        /// </summary>
        /// <returns></returns>
        public override string ToString()
        {
            return String.Join("&", (String[])_formData.ToArray(typeof(string)));
        }

        /// <summary>
        /// Identity Providers have different field names for credentials.   
        /// This abstraction allows the code to make one call to set credentials. 
        /// </summary>
        /// <param name="username"></param>
        /// <param name="password"></param>
        public abstract void SetCredentials(string username, string password);
    }
}

File: Java/VersionOneSSO/src/com/versionone/sso/PostData.java
package com.versionone.sso;

import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;
import java.util.ArrayList;
import java.util.List;

import org.apache.commons.lang.StringUtils;

/**
 * Data posted to SSO Providers (IDP and SP)
 * This object needs to be created and returned by the class that implements IResponseProvider
 * @author jerry
 *
 */
public abstract class PostData {

    private List<String> _formData = new ArrayList<String>();
    
    /**
     * Add a key-value pair
     * results in post data that read key=value
     * 
     * @param key
     * @param value
     */
    public void add(String key, String value)
    {
        try {
			_formData.add(String.format("%s=%s", key, URLEncoder.encode(value, "UTF-8")));
		} catch (UnsupportedEncodingException e) {
			e.printStackTrace();
			_formData.add(String.format("%s=%s", key, value));
		}
    }
    
    /**
     * adds a key with no value
     * results in post data that reads key=
     * @param key
     */
    public void add(String key)
    {
        _formData.add(String.format("%s=", key));
    }

    /**
     * Returns the post data string
     */
    @Override
	public String toString() {
    	return StringUtils.join(_formData, "&");
	}
    
    /**
     * Identity Providers have different field names for credentials.   
     * This abstraction allows the code to make one call to set credentials.  
     * @param username
     * @param password
     */
    public abstract void setCredentials(String username, String password);
}
